rfd/0145-api-client-credential-reload.md

```diff
@@ -0,0 +1,200 @@
+a change is made to the Teleport Client to support reloading credentials on a 
+fixed interval or when manually requested.
+
+For a developer writing a custom automation, it would look like so:
+
+```go
+func example() {
+    clt, err := client.New(ctx, client.Config{
+        Addrs: []string{
+        "tele.example.com:443",
+        },
+        Credentials: []client.Credentials{
+            client.LoadIdentityFile("my-identity-file"),
+        },
+		ReloadInterval: time.Minute * 5,
+    })
+	
+	// Or, rather than relying on ReloadInterval, ReloadCredentials() could be
+	// invoked in a file watcher.
+	err = clt.ReloadCredentials(ctx)
+}
+```
+
+The first way of implementing this would be to adjust the SSH based dialers to 
+support using the most recently reloaded `ssh.Config` when they are invoked 
+(rather than using the same `ssh.Config` throughout their lifetime) and would 
+also involve creating a custom `grpc/credentials.TransportCredentials` 
+implementation to support using the most recently reloaded `tls.Config`. This
+implementation would wrap over the existing TLS based `TransportCredential`
+offered by the package.
+
+This first implementation would provide similar behaviour to option A.
+
+Benefits:
+
+- Supports all credential loader types and wouldn't require modifications to
+  them.
+
+Drawbacks:
+
+- Much more "risky" than Option A without any significant benefits.
+- Changes in credentials (e.g roles associated with the identity) will not be
+  reflected in requests made to the server until the connection between the Auth
+  Server and client is broken. This could be anything from minutes to days
+  depending on the stability of the connection.
+
+The second way of implementing this would be for the Teleport API client to
+internally rotate the `grpc.ClientConn` being used. 
+
+Whilst this requires no changes to the credentials or the diallers, it does 
+involve managing the closure of the previous `grpc.ClientConn` - if this omitted
+then a memory leak will occur. Managing the closure of these could be complex as
+a streaming RPC could last a significant period of time after the client has 
+switched to the new `grpc.ClientConn` - and prematurely closing the 
+`grpc.ClientConn` will lead to the streaming RPC being interrupted. 
+
+An implementation could make use of a custom gRPC client balancer to manage a
+set of SubConns. A new `grpc.SubConn` could be created at rotation time and the
+balancer could begin to instruct the `grpc.ClientConn` to use this for RPCs.
+The balancer could then monitor the number of outstanding RPCs on the previous
+`grpc.SubConn` and close this when this becomes zero. Therefore not interrupting
+any existing RPCs.
+
+Benefits:
+
+- Changes in credentials can be reflected in requests immediately.
+- Supports all credential loader types and wouldn't require modifications to
+  them.
+
+Drawbacks:
+
+- A risky change to the core internals of our API client.
```

Comment:
the good part of using an LB is that you can enable it for some cases - `tbot` only LOL - so it won't be a major issue for the "rest" of teleport.
As part of engineering, yes, it involves much more work but one can control the connections and close them properly when no one uses them.

Reply:
Yeah although - it's not necessarily for `tbot` itself, really it's more about consumers of `tbot` in this case (so the events handler, log exporter, custom integrations)